### PR TITLE
Allow simulating calls in Account

### LINKED ISSUE #471: Account does not allow execution simulation in v0.4.0b
With Starknet v0.10.0, execution simulation has been introduced (`/feeder_gateway/simulate_transaction` of the sequencer API). It is used by `starknet invoke --estimate_fee` and `starknet invoke --simulate` of Starknet CLI.

This is how `starknet invoke` is implemented: [code](https://github.com/starkware-libs/cairo-lang/blob/v0.10.0/src/starkware/starknet/cli/starknet_cli.py#L967).
The linked function calls [this one](https://github.com/starkware-libs/cairo-lang/blob/v0.10.0/src/starkware/starknet/cli/starknet_cli.py#L565), which clearly sets `version` to `QUERY_VERSION` since `query=True`:

Should [this line of account.cairo](https://github.com/OpenZeppelin/cairo-contracts/blob/v0.4.0b/src/openzeppelin/account/library.cairo#L179) be changed to allow both `TRANSACTION_VERSION = 1` and `QUERY_VERSION = QUERY_BASE + 1`?

### PR DESCRIPTION
**Pull request by ericnordelo**
Fixes #471 

Allow transactions with version equal to the current Starknet QUERY_VERSION.

- [x] Tests
- [x] Tried the feature on a public network

### CODE REVIEW

**Comment by ericnordelo** [comment]
This version of the Account.cairo preset allows estimating fees or simulating transactions from starknet cli (or nile wrapping the latter), by accepting the transaction version for queries (2**128 + TRANSACTION_VERSION).

File: src/openzeppelin/utils/constants/library.cairo [hunk 1]
```diff
@@ -34,3 +34,11 @@ const IACCESSCONTROL_ID = 0x7965db0b;
...
 
 const DEFAULT_ADMIN_ROLE = 0;
+
+//
+// Starknet
+//
+
+const TRANSACTION_VERSION = 1;
+// For transactions meant to query and not to be executed by the Starknet OS
+const QUERY_VERSION = 2**128 + TRANSACTION_VERSION;
```
Comment (andrew-fleming) [inline src/openzeppelin/utils/constants/library.cairo:44]:
```suggestion
const QUERY_VERSION = 2**128 + TRANSACTION_VERSION;

```

Comment (andrew-fleming) [inline src/openzeppelin/utils/constants/library.cairo:44]: Do you think we should include another constant for `QUERY_VERSION_BASE` and insert that into `QUERY_VERSION` (like in the SW repo [here](https://github.com/starkware-libs/cairo-lang/blob/54d7e92a703b3b5a1e07e9389608178129946efc/src/starkware/starknet/definitions/constants.py#L55-L56))? It might add clarity about what `2**128` is doing.

File: src/openzeppelin/account/library.cairo [hunk 2]
```diff
@@ -175,13 +180,16 @@ namespace Account {
         alloc_locals;
 
         let (tx_info) = get_tx_info();
-        with_attr error_message("Account: invalid tx version") {
-            assert tx_info.version = 1;
+        // Allow query calls for simulations and assert transaction version
+        if (tx_info.version != QUERY_VERSION) {
+            with_attr error_message("Account: invalid tx version") {
+                assert tx_info.version = TRANSACTION_VERSION;
```
Comment (andrew-fleming) [inline src/openzeppelin/account/library.cairo:186]: Should we test that `QUERY_VERSION` txs will fail upon execution? I'm not sure that it's on us to do so, but I'm curious on what you think

**Review by andrew-fleming — COMMENTED** [review]
Looking good, @ericnordelo! I left a couple questions and one tiny suggestion. Also, do you think we should document the tx query feature in the Accounts docs?

File: src/openzeppelin/account/library.cairo (hunk 2 above)
Comment (ericnordelo) [reply src/openzeppelin/account/library.cairo:186]: That's protocol level. I don't think we need to test it.

File: src/openzeppelin/utils/constants/library.cairo (hunk 1 above)
Comment (ericnordelo) [reply src/openzeppelin/utils/constants/library.cairo:44]: Yes! I agree it adds clarity.

**Comment by andrew-fleming** [comment]
> > Looking good, @ericnordelo! I left a couple questions and one tiny suggestion. Also, do you think we should document the tx query feature in the Accounts docs?
> 
> I don't think we need to document that because that's a feature for the protocol itself, not for the end user.
> 
> This will be used transparently from starknet commands like fee estimation and transaction simulation, which should come with the corresponding documentation.

Yeahhhh I was leaning that way also. Sounds good, sir!

**Review by andrew-fleming — APPROVED** [review]
Looks good to me!

File: src/openzeppelin/utils/constants/library.cairo [hunk 3]
```diff
@@ -34,3 +34,13 @@ const IACCESSCONTROL_ID = 0x7965db0b;
...
 const DEFAULT_ADMIN_ROLE = 0;
+
+//
+// Starknet
+//
+
+const TRANSACTION_VERSION = 1;
+// The offset for query transaction versions
+const QUERY_VERSION_BASE = 2**128;
+// For transactions meant to query and not to be executed by the Starknet OS
```
Comment (martriay) [inline src/openzeppelin/utils/constants/library.cairo:45]:
```suggestion
// For transactions meant to simulate and not to be executed by the Starknet OS
```
I know "query" is the official word choice, but I find simulation/fee estimation more clear terms

File: src/openzeppelin/account/library.cairo (hunk 2 above)
Comment (martriay) [reply src/openzeppelin/account/library.cairo:186]: why would `QUERY_VERSION` fail on execution? this might be unrelated but while analyzing this change I realized these accounts are not future proof. the only reason we have the version check is to disallow txs of version 0.

if there were version 2 txs in the future, these accounts would stop working and require a migration, are we sure we want to impose this considering that accounts could be perfectly compatible with such txs?

if we choose to simply disallow version 0, I suggest we replace this double check (if + assert) with `assert_le` to make sure we allow any version > 0.

**Review by martriay — COMMENTED** [review]
Looking good, left a couple questions.

**Review by martriay — APPROVED** [review]

File: src/openzeppelin/utils/constants/library.cairo [hunk 4]
```diff
@@ -34,3 +34,13 @@ const IACCESSCONTROL_ID = 0x7965db0b;
...
+
+//
+// Starknet
+//
+
+const TRANSACTION_VERSION = 1;
+// The offset for query transaction versions
+const QUERY_VERSION_BASE = 2**128;
+// For transactions meant to simulate and not to be executed by the Starknet OS
+const QUERY_VERSION = QUERY_VERSION_BASE + TRANSACTION_VERSION;
```
Comment (martriay) [inline src/openzeppelin/utils/constants/library.cairo:46]: unsure about keeping these if they're not used

**Review by martriay — APPROVED** [review]

File: src/openzeppelin/account/library.cairo (hunk 2 above)
Comment (andrew-fleming) [reply src/openzeppelin/account/library.cairo]: > why would QUERY_VERSION fail on execution?

`QUERY_VERSION` txs should fail if they're not queries. 

> are we sure we want to impose this considering that accounts could be perfectly compatible with such txs?

Great question! It's definitely not future-proof and would require migration in future versions. IMO erring on the side of caution seems wise especially considering the gravitas of the account contract. We don't know (at least I don't) what future tx versions will look like. If tx version 2, for example, includes some new feature that requires another check, this could allow the possibility of using the version 1 model to bypass the version 2 check.

Maybe I'm being overly cautious. What do you guys think?

**Review by andrew-fleming — APPROVED** [review]
Looks good to go!

File: src/openzeppelin/account/library.cairo [hunk 5]
```diff
@@ -181,7 +182,7 @@ namespace Account {
         let (tx_info) = get_tx_info();
         // Disallow deprecated tx versions
         with_attr error_message("Account: deprecated tx version") {
-            assert_le(TRANSACTION_VERSION, tx_info.version);
+            assert is_le_felt(TRANSACTION_VERSION, tx_info.version) = TRUE;
```
Comment (martriay) [inline src/openzeppelin/account/library.cairo:185]: why was this needed?

**Review by martriay — APPROVED** [review]
